File: sdk/python/alp_sdk/cost_budget.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Dict, Any, Optional

class CostBudget:
    def __init__(
        self,
        budget_id: str,
        task_id: str,
        max_tokens: int,
        max_cost_usd: float,
        provider: str = "openai",
        model_tier: str = "standard",
        created_at: Optional[str] = None,
    ):
        self.id = budget_id
        self.task_id = task_id
        self.max_tokens = max_tokens
        self.max_cost_usd = max_cost_usd
        self.used_tokens = 0
        self.used_cost_usd = 0.0
        self.provider = provider
        self.model_tier = model_tier
        self.created_at = created_at or datetime.now(timezone.utc).isoformat()

class CostBudgetEngine:
    def __init__(self):
        self.budgets: Dict[str, CostBudget] = {}

    def create_budget(
        self,
        task_id: str,
        max_tokens: int,
        max_cost_usd: float,
        provider: str = "openai",
        model_tier: str = "standard",
    ) -> CostBudget:
        budget = CostBudget(
            budget_id=f"budget-{task_id}",
            task_id=task_id,
            max_tokens=max_tokens,
            max_cost_usd=max_cost_usd,
            provider=provider,
            model_tier=model_tier,
        )
        self.budgets[budget.id] = budget
        return budget

    def track_usage(
        self, budget_id: str, tokens_used: int, cost_usd: float
    ) -> Dict[str, Any]:
        budget = self.budgets.get(budget_id)
        if not budget:
            return {"remaining_cost_usd": 0.0, "remaining_tokens": 0, "is_exceeded": True}

        budget.used_tokens += tokens_used
        budget.used_cost_usd += cost_usd

        rem_cost = max(0.0, budget.max_cost_usd - budget.used_cost_usd)
        rem_tokens = max(0, budget.max_tokens - budget.used_tokens)
        is_exceeded = budget.used_cost_usd > budget.max_cost_usd or budget.used_tokens > budget.max_tokens

        return {
            "remaining_cost_usd": rem_cost,
            "remaining_tokens": rem_tokens,
            "is_exceeded": is_exceeded,
        }

    def select_optimal_model(
        self, task_complexity: str, max_cost_usd: float
    ) -> Dict[str, Any]:
        if task_complexity == "high" and max_cost_usd >= 0.10:
            return {"provider": "anthropic", "model": "claude-3-5-sonnet", "estimated_cost_per_1k": 0.003}
        elif task_complexity == "medium" or max_cost_usd >= 0.02:
            return {"provider": "openai", "model": "gpt-4o-mini", "estimated_cost_per_1k": 0.00015}
        else:
            return {"provider": "ollama", "model": "llama3.2-local", "estimated_cost_per_1k": 0.00}

File: parser/src/cost-budget.ts
export interface CostBudget {
  id: string;
  taskId: string;
  maxTokens: number;
  maxCostUSD: number;
  usedTokens: number;
  usedCostUSD: number;
  provider: string;
  modelTier: 'standard' | 'premium' | 'economy';
  createdAt: string;
}

export class CostBudgetEngine {
  private budgets: Map<string, CostBudget> = new Map();

  public createBudget(
    taskId: string,
    maxTokens: number,
    maxCostUSD: number,
    provider: string = 'openai',
    modelTier: 'standard' | 'premium' | 'economy' = 'standard'
  ): CostBudget {
    const budget: CostBudget = {
      id: `budget-${taskId}`,
      taskId,
      maxTokens,
      maxCostUSD,
      usedTokens: 0,
      usedCostUSD: 0,
      provider,
      modelTier,
      createdAt: new Date().toISOString(),
    };

    this.budgets.set(budget.id, budget);
    return budget;
  }

  public trackUsage(budgetId: string, tokensUsed: number, costUSD: number): { remainingCostUSD: number; remainingTokens: number; isExceeded: boolean } {
    const budget = this.budgets.get(budgetId);
    if (!budget) {
      return { remainingCostUSD: 0, remainingTokens: 0, isExceeded: true };
    }

    budget.usedTokens += tokensUsed;
    budget.usedCostUSD += costUSD;

    const remainingCostUSD = Math.max(0, budget.maxCostUSD - budget.usedCostUSD);
    const remainingTokens = Math.max(0, budget.maxTokens - budget.usedTokens);
    const isExceeded = budget.usedCostUSD > budget.maxCostUSD || budget.usedTokens > budget.maxTokens;

    return {
      remainingCostUSD,
      remainingTokens,
      isExceeded,
    };
  }

  public selectOptimalModel(taskComplexity: 'low' | 'medium' | 'high', maxCostUSD: number): { provider: string; model: string; estimatedCostPer1k: number } {
    if (taskComplexity === 'high' && maxCostUSD >= 0.10) {
      return { provider: 'anthropic', model: 'claude-3-5-sonnet', estimatedCostPer1k: 0.003 };
    } else if (taskComplexity === 'medium' || maxCostUSD >= 0.02) {
      return { provider: 'openai', model: 'gpt-4o-mini', estimatedCostPer1k: 0.00015 };
    } else {
      return { provider: 'ollama', model: 'llama3.2-local', estimatedCostPer1k: 0.00 };
    }
  }

  public getBudget(id: string): CostBudget | undefined {
    return this.budgets.get(id);
  }
}
